### PR TITLE
fix 'Set Linear' not saving properly (#1642)

### LINKED ISSUE #1642: "Set Linear" not saving properly
**Edited 2015-01-16 @tresf**

The `Set Linear` function in some plugins doesn't seem to be saving properly.  Steps to reproduce:

 1. Open a new project
 2. Add `Calf Reverb` effect to master channel
 3. Change `Bass Cut` knob to `Set linear`
 4. Save project.
 5. Reopen project, examine `Bass Cut` `Set linear` knob settings was not saved

**Note:**  If this is done on both `Bass Cut` as well as `Treble Cut`, the `Set linear` setting will be saved Bass Cut, not Treble Cut.

-Tres

---

**Original bug report**

Hi guys

I am not sure if the issue I had is the same, but to me it looks like it.

I had made a song with the previous version. After loading it in lmms, I had issues with two automation tracks. The sound was distorted. Redoing those automation tracks worked for me.
On the lmms forum however there is a topic with the same title as this one.
Some one is having an issue with the calf filter.
http://lmms.io/forum/viewtopic.php?f=7&t=1557

### PR DESCRIPTION
**Pull request by serdnab**
I added `setScaleType( Linear )` when the loading element doesn't have a child element (containing scale type or automation data) as this is the default save way for non automated linear knobs
fixes #1642

### CODE REVIEW

**Comment by Umcaruje** [comment]
@serdnab could you please rename the PR to something more descriptive like " Fix 'Set Linear' not saving properly"

**Comment by Umcaruje** [comment]
Did a quick test, seems to fix #1642 for me. I'm not comfortable with merging this though, without any more extensive testing, because I'm not familiar with this part of the codebase at all.

**Comment by Umcaruje** [comment]
@LMMS/developers could someone else review/test this PR out, so we could merge?

**Comment by jasp00** [comment]
Should not the backward compatibility code be in an upgrade method in `src/core/DataFile.cpp`?

**Comment by Umcaruje** [comment]
@serdnab could you please address the issue @jasp00 pointed out?

**Comment by serdnab** [comment]
Which is your opinion on opening old projects that their origin versions of lmms didn't have logarithmic knobs. Open it linear or as detected by `isLogarithmic()` in `LadspaManager.cpp`?

**Comment by tresf** [comment]
> Which is your opinion on opening old projects that their origin versions of lmms didn't have logarithmic knobs. Open it linear or as detected by isLogarithmic() in LadspaManager.cpp?

Linear.

**Comment by tresf** [comment]
Also as @jasp00 points out, we want to always preserve the bug when possible.  e.g. if it was incorrectly setting Logarithmic before when Linear was explicitly being clicked, write `Logarithmic` to the config and make them change it again to take effect.  These things can really destroy projects if not upgraded properly.

**Comment by serdnab** [comment]
@Umcaruje 

> could you please address the issue @jasp00 pointed out?

done

**Comment by jasp00** [comment]
@serdnab: This branch has conflicts that must be resolved

**Comment by serdnab** [comment]
@jasp00 

> This branch has conflicts that must be resolved

solved

File: include/DataFile.h [hunk 1]
```diff
@@ -122,14 +123,17 @@ class EXPORT DataFile : public QDomDocument
...
 	void upgrade_0_4_0_rc2();
+	void upgrade_1_0_99();
 	void upgrade_1_1_0();
 	void upgrade_1_1_91();
 
 	void upgrade();
 
 	void loadData( const QByteArray & _data, const QString & _sourceFile );
 
+	void findIds(const QDomElement& elem, QList<jo_id_t>& idList);
```
Comment (jasp00) [inline include/DataFile.h]: This function does not depend on the class. It should be a static function in `DataFile.cpp`.

File: src/core/DataFile.cpp [hunk 2]
```diff
@@ -794,6 +794,45 @@ void DataFile::upgrade_0_4_0_rc2()
...
+			{
+				if (el.attribute("scale_type") == "log")
+				{
+					QDomElement me = createElement("data");
+					me.setAttribute("value", el.attribute("data"));
+					me.setAttribute("scale_type", "log");
+					
+					jo_id_t id;
+					for(jo_id_t tid = rand(); idList.contains(id = tid % EO_ID_MSB
+															  | EO_ID_MSB); tid++)
```
Comment (jasp00) [inline src/core/DataFile.cpp]: The ids from loaded projects do not have `EO_ID_MSB` set, that would conflict with new objects. This id allocation from `ProjectJournal` should not be here; just assign any free id, that is, `tid = last_assigned + 1` starting from 0.

Assume tabs are 8 spaces. This line is difficult to see in the web interface.

**Comment by serdnab** [comment]
@jasp00 
addressed your requests

File: src/core/DataFile.cpp [hunk 3]
```diff
@@ -794,6 +797,47 @@ void DataFile::upgrade_0_4_0_rc2()
...
+			{
+				if (el.attribute("scale_type") == "log")
+				{
+					QDomElement me = createElement("data");
+					me.setAttribute("value", el.attribute("data"));
+					me.setAttribute("scale_type", "log");
+					
+					jo_id_t id;
+					for(jo_id_t tid = last_assigned_id + 1;
+						idList.contains(id = tid); tid++)
```
Comment (jasp00) [inline src/core/DataFile.cpp]: `id` should be used directly in the loop, `tid` is not needed.

**Comment by serdnab** [comment]
@jasp00 

done
deleted the unnecessary variable

File: src/core/DataFile.cpp [hunk 4]
```diff
@@ -794,6 +797,47 @@ void DataFile::upgrade_0_4_0_rc2()
 }
 
 
+void DataFile::upgrade_1_0_99()
+{
+	jo_id_t last_assigned_id = 0;
+
```
Comment (jasp00) [inline src/core/DataFile.cpp]: Empty lines should not have any indentation, but coding conventions are not against this.

**Comment by jasp00** [comment]
Current projects do not need a manual upgrade in this case.

**Pull request merged**